dhcp/dhcp.go

```diff
@@ -0,0 +1,365 @@
+	packet.Write(make([]byte, bytesInAddress))
+
+	// chaddr: Client hardware address (MAC address)
+	paddingBytes := 10
+	packet.Write(mac)                        // MAC address (6 bytes)
+	packet.Write(make([]byte, paddingBytes)) // Padding to 16 bytes
+
+	// sname: Server host name (64 bytes)
+	serverHostNameBytes := 64
+	packet.Write(make([]byte, serverHostNameBytes))
+	// file: Boot file name (128 bytes)
+	bootFileNameBytes := 128
+	packet.Write(make([]byte, bootFileNameBytes))
+
+	// Magic cookie (DHCP)
+	err = binary.Write(&packet, binary.BigEndian, magicCookie)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to write magic cookie")
+	}
+
+	// DHCP options (minimal required options for DISCOVER)
+	packet.Write([]byte{
+		53, 1, 1, // Option 53: DHCP Message Type (1 = DHCP Discover)
+		55, 3, 1, 3, 6, // Option 55: Parameter Request List (1 = Subnet Mask, 3 = Router, 6 = DNS)
+		255, // End option
+	})
+
+	// padding length to 300 bytes
+	var value uint8 // default is zero
+	if packet.Len() < bootpMinLen {
+		packet.Write(bytes.Repeat([]byte{value}, bootpMinLen-packet.Len()))
+	}
+
+	return packet.Bytes(), nil
+}
+
+// MakeRawUDPPacket converts a payload (a serialized packet) into a
+// raw UDP packet for the specified serverAddr from the specified clientAddr.
+func MakeRawUDPPacket(payload []byte, serverAddr, clientAddr net.UDPAddr) ([]byte, error) {
+	udpBytes := 8
+	udp := make([]byte, udpBytes)
+	binary.BigEndian.PutUint16(udp[:2], uint16(clientAddr.Port))
+	binary.BigEndian.PutUint16(udp[2:4], uint16(serverAddr.Port))
+	totalLen := uint16(udpBytes + len(payload))
+	binary.BigEndian.PutUint16(udp[4:6], totalLen)
+	binary.BigEndian.PutUint16(udp[6:8], 0) // try to offload the checksum
+
+	headerVersion := 4
+	headerLen := 20
+	headerTTL := 64
+
+	h := ipv4.Header{
+		Version:  headerVersion, // nolint
+		Len:      headerLen,     // nolint
+		TotalLen: headerLen + len(udp) + len(payload),
+		TTL:      headerTTL,
+		Protocol: udpProtocol, // UDP
+		Dst:      serverAddr.IP,
+		Src:      clientAddr.IP,
+	}
+	ret, err := h.Marshal()
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to marshal when making udp packet")
+	}
+	ret = append(ret, udp...)
+	ret = append(ret, payload...)
+	return ret, nil
+}
+
+// Send DHCP discover packet using unix.RawConn
+func (c *DHCP) sendDHCPDiscover(fd int, packet []byte) error {
+	raddr := &net.UDPAddr{IP: net.IPv4bcast, Port: dhcpServerPort}
+	laddr := &net.UDPAddr{IP: net.IPv4zero, Port: dhcpClientPort}
+	var destination [net.IPv4len]byte
+	copy(destination[:], raddr.IP.To4())
+
+	packetBytes, err := MakeRawUDPPacket(packet, *raddr, *laddr)
+	if err != nil {
+		return errors.Wrap(err, "error making raw udp packet")
+	}
+
+	// Create sockaddr_ll structure for sending the packet
+	remoteAddr := unix.SockaddrInet4{Port: laddr.Port, Addr: destination}
+
+	// Send the packet using the raw socket
+	return errors.Wrap(unix.Sendto(fd, packetBytes, 0, &remoteAddr), "failed unix send to")
```

Comment:
Passing around ints for file descriptors is a bit too C-like. The Go approach would be to create an `io.Writer` that implements `Write()` using `unix.Sendto`. That writer then gets passed around (as an `io.Writer`). That would also extract the linux-specific parts to the Writer implementation and leave the packet manufacturing platform-independent.

That might help with testing too, since `io.Writers` can be as simple as `bytes.Buffer{}`.